keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,388 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The main goal of this enhancement is to authoritatively identify the Operating System
+on which a pod would like to run.
+
+
+## Motivation
+
+Identifying the OS of the pods during the API Server admission time is crucial to apply
+appropriate security constraints to the pod. Without proper identification, some admission
+plugins may apply unnecessary security constraints to the pods or in the
+worst case, don't apply security constraints at all.
+
+
+### Goals
+
+Admission plugins and/or Admission Webhooks identifying the Operating System on which
+pod intends to run.
+
+### Non-Goals
+
+
+- Validating if the pods have appropriate security constraints.
+- Interaction with container runtimes.
+
+## Proposal
+
+
+Identifying the Windows pods during the scheduling phase can be done
+in the following ways:
+- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
+  Windows node specific labels.
+- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec
+
+The runtimeclass is a higher level abstraction which gets translated again to nodeSelectors+tolerations. While the 
+nodeSelector with reserved OS label is good enough, it can be viewed as an side-effect of scheduling constraints 
+and need not necessarily express the user intention fully. So, we propose to add a new field to the pod spec called 
+os to identify the OS on which pod intends to run.
+
+```go
+type PodSpec struct {
+  // If specified, indicates the type of OS on which pod intends
+  // to run.  This field is alpha-level and is only honored by servers that 
+  // enable the IdentifyOS feature. This is used to help identifying the
+  // OS authoritatively during the API server admission.
+  // +optional
+  OS *OS
+}
+
+// OS has information on the type of OS. 
+// We're making this a struct for possible future expansion.
+type OS struct {
```

Comment:
There was a fair amount of discussion related to minor versions of windows as specific targets.  As an API consumer, I'd really like to assume all my windows pods have the same fields valid, but I don't actually know.  @liggitt?

Reply:
I'm OK with treating all versions of Windows as the same for admission/validation purposes.
Almost all functionality that is only supported on some versions of Windows is in networking (like IPv4/IPv6 dualstack support) and these behaviors are usually controlled by kube-proxy flags or Service specs not pod specs.
@jsturtevant @jayunit100 @claudiubelu do any of you have concerns here?

Reply:
I am resolving the comment here. @deads2k if your comment or question is unanswered please let me know.